README.md
# Tree Tracker Web Map (BackEnd - API)

This project is built using Node.js version 12.19.0.

Project Knowledge Base: https://app.gitbook.com/@greenstand/s/impact-map/

## Requirements

- Node.js 12

## News

The docker image name was updated from treetracker-webmap-api to treetracker-web-map-api due to standardization in how we name docker images based of repository names.  This change had introduced deployment issues.  Cloud engineers please be adivsed of this change.

## Backend Development

To set up your environment for BackEnd development, follow the steps below:

> NOTE: We recommend all of these tasks to be executed using Visual Studio Code, openning a Terminal session.


1. Install node modules, running npm i in the root
    ```
    npm i
    ```
2. Set up database connection string.  

    
    1. Create a .env file at the folder server.
    2. Add the following content in the .env file

        ```
        DATABASE_URL=postgresql://xxxx:xxxx@dxxxx:25060/treetracker?ssl=no-verify
        ```

        > NOTE: The default mode to run the API in the test environment is connecting with the remote database.
            > * Ask for the db connectiong string on the slack channel.
    

3. Install supervisor globally

    ```
    npm i -g supervisor
    ```

4. <a name="rundev">Run the Backend Development Environment</a>

    Attention:

    - Option 1 is for **BackEnd development only**.
    - Option 2 is for **FrontEnd and BackEnd development**.

    Choose one of the options and      Run the code below on the terminal:

    - **Option 1**:

         > NOTE 1: This command will run server.js with CORS restrictions lifted
         > NOTE 2: For environments running the FrontEnd and BackEnd, do not use this option, because API will run at port 3000 conflicting with the React project. Choose option 2.
       

        ```
        NODE_ENV=dev supervisor server.js
        ```
        

    - **Option 2**: 

        ```
        cd server
        npm run dev
        ```
        > NOTE: 
        > 1. In this way, the server would run on port 3001, and client would run on port 3000.

<br>  

> IMPORTANT NOTE: When running for the first time the API, if the endpoints are not responding, because after the queries there aren't any answer or error, it might be some incompatibility with the node version and pg module. If you are running Node.js 14, downgrade to Node.js 12.19.0.
[How to Upgrade (or Downgrade) Node.js Using npm](https://www.surrealcms.com/blog/how-to-upgrade-or-downgrade-nodejs-using-npm.html)

<br>  



### (Optional) Local Database



* The steps below are required only for scenarios where the API needs to connect with the Postgres database locally.
* The default mode to run the application is connecting with the remote database, but for some specific scenarios, as an exception, the developer might need to have the database running local.
* Do not execute these steps if you are configuring the environment for the first time.

1. Install postgres
2. Install postGIS
3. Create a database named 'treetracker'
4. Import our developer seed into your database.  
    - Ask in slack for the link to the seed.
5. Configure .env to point to your local database.
6. Run the [Backend Development Environmet](#rundev).


---


### How to test

We use Jest to build tests.

1. To test server
```
cd server
npm test
```
Troubleshooting Local Set up - If your local set up is showing errors , comment the line in (engine-strict=true) .npmrc file.


CHANGELOG.md
# [1.17.0](https://github.com/Greenstand/treetracker-web-map/compare/v1.16.1...v1.17.0) (2022-11-09)


### Features

* planting_org ([929561f](https://github.com/Greenstand/treetracker-web-map/commit/929561feb28fc4b317c5b5f2e5427ff86bbe0ac3))

## [1.16.1](https://github.com/Greenstand/treetracker-web-map/compare/v1.16.0...v1.16.1) (2022-02-28)


### Bug Fixes

* get tree by token ([dbf0319](https://github.com/Greenstand/treetracker-web-map/commit/dbf0319a36f2f0e5ed7c015f3817a76d3734d50f))
* release node from 12 -> 14 ([2bcb88e](https://github.com/Greenstand/treetracker-web-map/commit/2bcb88e77c6f3bcb4897161e736bac2222fa8f1b))
* remove extra line ([862a4b5](https://github.com/Greenstand/treetracker-web-map/commit/862a4b59994e5ebe7d26c13de1fad0269208f837))
* remove useless code ([88c4ef0](https://github.com/Greenstand/treetracker-web-map/commit/88c4ef00471ef901f2b0488a37ef2e57007ad40c))

# [1.16.0](https://github.com/Greenstand/treetracker-web-map/compare/v1.15.1...v1.16.0) (2021-08-14)


### Features

* wallet switch to case1 when big ([2c54bb2](https://github.com/Greenstand/treetracker-web-map/commit/2c54bb2a64176229fae19182ca05cccff9dd40f0))

## [1.15.1](https://github.com/Greenstand/treetracker-web-map/compare/v1.15.0...v1.15.1) (2021-08-05)


### Bug Fixes

* use delete to remove key ([94ad2d8](https://github.com/Greenstand/treetracker-web-map/commit/94ad2d8eed2d974a4955fd8feea94c22464af464))

# [1.15.0](https://github.com/Greenstand/treetracker-web-map/compare/v1.14.2...v1.15.0) (2021-08-05)


### Features

* remove planter identifer from tree detail object ([af0949b](https://github.com/Greenstand/treetracker-web-map/commit/af0949bf1301f39cd252b0523ec972b948bed225))

## [1.14.2](https://github.com/Greenstand/treetracker-web-map/compare/v1.14.1...v1.14.2) (2021-06-11)


### Bug Fixes

* return undefine rather than throw error when didnt' find tree ([002008d](https://github.com/Greenstand/treetracker-web-map/commit/002008dad0fc1c2e183ce4044695de8c4dcd08dd))

## [1.14.1](https://github.com/Greenstand/treetracker-web-map/compare/v1.14.0...v1.14.1) (2021-06-10)


### Bug Fixes

* small bug on printing log ([8bb6035](https://github.com/Greenstand/treetracker-web-map/commit/8bb60358b557786f2e89482f53e04a683f3b7963))

# [1.14.0](https://github.com/Greenstand/treetracker-web-map/compare/v1.13.1...v1.14.0) (2021-06-10)


### Features

* support api: /tree?tree_name=xxx ([ee16e63](https://github.com/Greenstand/treetracker-web-map/commit/ee16e63eb9fa8a2618b7fc989d0d2fa0db3f3f4b))

## [1.13.1](https://github.com/Greenstand/treetracker-web-map/compare/v1.13.0...v1.13.1) (2021-05-22)


### Bug Fixes

* remove old workflow ([839dc53](https://github.com/Greenstand/treetracker-web-map/commit/839dc5397d07bceeea05d1626a5c259ecdc118d8))

# [1.13.0](https://github.com/Greenstand/treetracker-web-map/compare/v1.12.0...v1.13.0) (2021-05-22)


### Features

* add note about docker image name change ([39f979b](https://github.com/Greenstand/treetracker-web-map/commit/39f979b43ce97268ae842a15baee2a69e2702407))

# [1.12.0](https://github.com/Greenstand/treetracker-web-map/compare/v1.11.9...v1.12.0) (2021-05-19)


### Features

* implement tree_name feature ([c1b2330](https://github.com/Greenstand/treetracker-web-map/commit/c1b2330f2a0d682589eec01d96b9eadbd918a7a1))

## [1.11.9](https://github.com/Greenstand/treetracker-web-map/compare/v1.11.8...v1.11.9) (2021-05-04)


### Bug Fixes

* workflow had extra steps in wrong order ([ff23c6c](https://github.com/Greenstand/treetracker-web-map/commit/ff23c6c2b917f017bed80f514f744253a6de8b7d))

## [1.11.8](https://github.com/Greenstand/treetracker-web-map/compare/v1.11.7...v1.11.8) (2021-05-04)


### Bug Fixes

* add eslint config ([975788b](https://github.com/Greenstand/treetracker-web-map/commit/975788b4d8857da9102c437773f632f7245c5dae))
* bring package-lock.json up to date ([ea39392](https://github.com/Greenstand/treetracker-web-map/commit/ea39392a2f2f1bcf64c3883a8fac0e6da8a9f171))
* skip tests ([293af67](https://github.com/Greenstand/treetracker-web-map/commit/293af67d7ec041686d86566dac380a72014d57a5))

## [1.11.7](https://github.com/Greenstand/treetracker-web-map/compare/v1.11.6...v1.11.7) (2021-04-12)


### Bug Fixes

* temporarily open CORS ([55dc487](https://github.com/Greenstand/treetracker-web-map/commit/55dc4874c6192eef40d656165a04caa6ba652eaf))


client/js/map.js
const YES = "YES";
const NO = "NO";
var map = undefined; //Google Map object
var mc = undefined; //Marker Clusterer
var markers = []; //All the markers
var mapModel = undefined;
var token;
var organization;
var treeid;
var userid;
var donor;
var wallet;
var flavor;
var clusterRadius;
var firstQuery = true;
var firstRender = true;
var firstInteraction = false;
var initialBounds = new google.maps.LatLngBounds();
var loader;
var panelLoader;

var currentZoom;
var req = null;
var treeInfoDivShowing = false;
var fetchMarkers = true;

// used to keep track of our points and markers
var points = [];
var markerByPointId = {};
var selectedTreeMarker;
var selectedOldTreeMarker;

//var treetrackerApiUrl = "http://dev.treetracker.org/api/web/";
var treetrackerApiUrl = "/api/web/";

if (typeof configTreetrackerApi !== "undefined") {
  treetrackerApiUrl = configTreetrackerApi;
}

/**
 * Writes to the document's cookie to expire (or not) after a set time from the date of the user's last visit to the webpage.
 *
 * @param {string=} c_name The name of the cookie to be written (i.e. "visited").
 * @param {string=} value The value of the cookie (i.e. "yes")
 * @param {Number=} exdays The length of time that the cookie will not expire
 */
function setCookie(c_name,value,exdays){
  var exdate=new Date();
  exdate.setDate(exdate.getDate() + exdays);
  var c_value=escape(value) + ((exdays==null) ? "" : "; expires="+exdate.toUTCString());
  document.cookie=c_name + "=" + c_value;
}

/**
 * Reads from the cookie if it exists.
 *
 * @param {string=} c_name The name of the cookie to be read from (i.e. "visited").
 */
function getCookie(c_name){
  var c_value = document.cookie;
  var c_start = c_value.indexOf(" " + c_name + "=");
  if (c_start == -1){
    c_start = c_value.indexOf(c_name + "=");
  }

  if (c_start == -1){
    c_value = null;
  } else{
    c_start = c_value.indexOf("=", c_start) + 1;
    var c_end = c_value.indexOf(";", c_start);

    if (c_end == -1){
      c_end = c_value.length;
    }
    c_value = unescape(c_value.substring(c_start,c_end));
  }
  return c_value;
}

/**
 * Checks if the user has visited the page before.
 *
 */
function checkSession(){
  var c = getCookie("visited");
  var past_visitor = true;
  if (c === "yes") {
    console.log("OLD VISITOR");
  } else {
    console.log("NEW VISITOR");
    past_visitor = false;
  }
  setCookie("visited", "yes", 365); // expire in 1 year; or use null to never expire
  return past_visitor;
}

function getTreeQueryParametersFromRequestedFilters(){
  var queryUrl = "";
  if (token != null) {
    queryUrl = queryUrl + "&token=" + token;
  } else if (organization != null) {
    queryUrl = queryUrl + "&organization=" + organization;
  } else if (treeid != null) {
    queryUrl = queryUrl + "&treeid=" + treeid;
  } else if (userid != null) {
    queryUrl = queryUrl + "&userid=" + userid;
  } else if (flavor != null) {
    queryUrl = queryUrl + "&flavor=" + flavor;
  } else if (wallet != null) {
    queryUrl = queryUrl + "&wallet=" + wallet;
  }
  return queryUrl;
}

//Get the tree data and create markers with corresponding data
var initMarkers = function(viewportBounds, zoomLevel) {
  console.log("initMarkers:", viewportBounds, zoomLevel);
  // no need to load this up at every tiny movement
  if (!fetchMarkers) {
    console.log("initMarkers quit");
    return;
  }

  clusterRadius =
    getQueryStringValue("clusterRadius") || getClusterRadius(zoomLevel);

  console.log("Cluster radius: " + clusterRadius);
  if (req != null) {
    console.log("initMarkers abort");
    req.abort();
  }
  var queryUrl = treetrackerApiUrl + "trees?clusterRadius=" + clusterRadius;
  queryUrl = queryUrl + "&zoom_level=" + zoomLevel;
  if (
    currentZoom >= 4 &&
    !(
      (token != null ||
        organization != null ||
        treeid != null ||
        userid !== null) &&
      firstRender == true
    )
  ) {
    queryUrl = queryUrl + "&bounds=" + viewportBounds;
  }
  queryUrl = queryUrl + getTreeQueryParametersFromRequestedFilters();

  console.log("request:", queryUrl);
  req = $.get(queryUrl, function(data) {
    console.log("initMarkers, get");
    if (userid && data.data.length === 0) {
      showAlert();
    }

    // clear everything
    points = [];
    markerByPointId = {};
    clearOverlays(markers);
    // console.log(data);

    $.each(data.data, function(i, item) {
      if (item.type == "cluster") {
        var centroid = JSON.parse(item.centroid);
        var latLng = new google.maps.LatLng(
          centroid.coordinates[1],
          centroid.coordinates[0]
        );
        determineInitialSize(latLng);

        var iconUrl = null,
          labelOrigin = null,
          anchor = null;
        if (item.count <= 300) {
          iconUrl = "/img/cluster_46px.png";
          labelOrigin = new google.maps.Point(23, 23);
          anchor = new google.maps.Point(23, 23);
        } else {
          iconUrl = "/img/cluster_63px.png";
          labelOrigin = new google.maps.Point(32, 32);
          anchor = new google.maps.Point(32, 32);
        }

        var marker = new google.maps.Marker({
          position: latLng,
          map: map,
          label: {
            text: shortenLargeNumber(item.count).toString(),
            color: "#000"
          },
          icon: {
            url: iconUrl,
            labelOrigin: labelOrigin,
            anchor: anchor
          }
        });

        //add zoomTarget to cluster marker
        marker.zoomTarget = data.zoomTargets?
          data.zoomTargets.reduce((a,c) => {
            if(a){
              return a;
            }else{
              if(c.region_id === item.id){
                return c;
              }else{
                return a;
              }
            }
          }, undefined)
        :
          undefined;

        google.maps.event.addListener(marker, "click", function() {
          if(marker.zoomTarget){
            fetchMarkers = false;
            var zoomLevel = map.getZoom();
            map.setZoom(zoomLevel + 2);
            const centroid = JSON.parse(marker.zoomTarget.centroid);
            const position = {
              lat: centroid.coordinates[1],
              lng: centroid.coordinates[0],
            }
            console.log("zoom target:", position);
            map.panTo(position);
          }else{
            fetchMarkers = false;
            var zoomLevel = map.getZoom();
            map.setZoom(zoomLevel + 2);
            map.panTo(marker.position);
          }
        });
        markers.push(marker);
      } else if (item.type == "point") {
        var latLng = new google.maps.LatLng(item.lat, item.lon);
        determineInitialSize(latLng);
        var infowindow = new google.maps.InfoWindow({
          content: "/img/loading.gif"
        });

        var marker = new google.maps.Marker({
          position: latLng,
          map: map,
          title: "Tree",
          icon: {
            url: "/img/pin_29px.png"
          },
          zIndex: undefined,
          payload: {
            id: item["id"]
          }
        });

        if (
          selectedTreeMarker &&
          marker.payload.id === selectedTreeMarker.payload.id
        ) {
          selectedTreeMarker = marker;
          changeTreeMarkSelected();
        }

        // set the field for sorting
        item._sort_field = new Date(item.time_created);

        // hold the reference to our points
        points.push(item);
        markerByPointId[item["id"]] = marker;
        markers.push(marker);
      }
    });

    // set the markers once we are done
    setPointMarkerListeners();

    if (firstRender) {
      // create infowindow object
      var infowindow = new google.maps.InfoWindow({
        content: "<div style='float:left'><img src='/img/TipPopupIcon.png' height=40 width=40></div><div style='float:right; padding: 10px;'><b>Click on the cluster to zoom into trees</b></div>"
      });
      //
      if (!checkSession()) { //only if the user is new
        // add the infowindow to a random starting marker to be visible by default when the user first loads the screen
        infowindow.open(map, markers[Math.floor(Math.random() * markers.length)]);
      }

      loader.classList.remove("active");
      firstRender = false;
    }
    console.log("init markert finished, loaded:", markers.length);
    //debugger;
    mapModel.checkArrow();
  });
};

// for each point, set the listeners.
// sort first so we can reference the next point
// in chronological order
function setPointMarkerListeners() {
  // points.sort(function(a, b) {
  //     return a._sort_field - b._sort_field;
  // });

  panelLoader = document.getElementById("tree-info-loader");

  $.each(points, function(i, point) {
    var marker = markerByPointId[point.id];
    google.maps.event.addListener(marker, "click", function() {
      panelLoader.classList.add("active");
      showMarkerInfo(point, marker, i);
      $("#tree-image").on("load", function() {
        panelLoader.classList.remove("active");
      });
    });
  });
}

function showAlert() {
  const alertHtml = `
    <div class="alert alert-info alert-dismissible" role="alert">
      <button type="button" class="close" data-dismiss="alert" aria-label="Close">
        <span aria-hidden="true">&times;</span>
      </button>
      Could not find any trees associated with userid ${userid}
    </div>
  `;
  // Prevent duplicate alerts after map is re-rendered
  if ($(".alert-container").find(".alert").length === 0) {
    $(alertHtml).prependTo(".alert-container");
  }
}

// set up and show the marker info
function showMarkerInfo(point, marker, index) {
  panelLoader = document.getElementById("tree-info-loader");

  $("#tree_info_div").show("slide", "swing", 600);
  if (treeInfoDivShowing == false) {
    treeInfoDivShowing = true;
    if (
      typeof window.orientation !== "undefined" ||
      navigator.userAgent.indexOf("IEMobile") !== -1
    ) {
      $("#map-canvas").animate(
        {
          margin: "0 0 0 20vw"
        },
        700,
        function() {
          //Animation Complete
        }
      );
    } else {
      $("#map-canvas").animate(
        {
          margin: "0 0 0 354px"
        },
        700,
        function() {
          //Animation Complete
        }
      );
    }
  }

  //toggle tree mark
  selectedOldTreeMarker = selectedTreeMarker;
  selectedTreeMarker = marker;
  changeTreeMarkSelected();

  // always center this one
  map.panTo(marker.getPosition());

  $("#create-data").html(
    moment(point["time_created"]).format("MM/DD/YYYY hh:mm A")
  );
  if (wallet != null) {
    $("#created_on").hide();
    $("#tree_id_holder").hide();
    $("#impact-owner-data").html("@" + wallet);
    $("#status-data").html("Token issued");
    $("#token-id-data").html(point["token_uuid"]);
  } else {
    $("#sponsor").hide();
    $("#token_holder").hide();
  }
  $("#updated-data").html(point["time_updated"]);
  $("#gps-accuracy-data").html(point["gps_accuracy"]);
  $("#latitude-data").html(point["lat"]);
  $("#longitude-data").html(point["lon"]);
  if (point["missing"]) {
    $("#missing-data").html(YES);
  } else {
    $("#missing-data").html(NO);
  }
  if (point["dead"]) {
    $("#dead-data").html(YES);
  } else {
    $("#dead-data").html(NO);
  }
  $("#tree-image").attr("src", point["image_url"]);
  $("#tree-id").html(point["id"]);
  $("#planter_name").html(
    point["first_name"] + " " + point["last_name"].slice(0, 1)
  );
  if (point["user_image_url"]) {
    $("#planter_image").attr("src", point["user_image_url"]);
  } else {
    $("#planter_image").attr("src", "/img/LogoOnly_Bright_Green100x100.png");
  }
  $("#tree_next").val(getCircularPointIndex(index + 1));
  $("#tree_prev").val(getCircularPointIndex(index - 1));

  $("#tree_next")
    .off("click")
    .on("click", function() {
      fetchMarkers = false;
      var index = parseInt($(this).val(), 10);
      panelLoader.classList.add("active");
      showMarkerInfoByIndex(index);
      $("#tree-image").on("load", function() {
        panelLoader.classList.remove("active");
      });
    });

  $("#tree_prev")
    .off("click")
    .on("click", function() {
      fetchMarkers = false;
      var index = parseInt($(this).val(), 10);
      panelLoader.classList.add("active");
      showMarkerInfoByIndex(index);
      $("#tree-image").on("load", function() {
        panelLoader.classList.remove("active");
      });
    });
}

function changeTreeMarkSelected() {
  if (selectedOldTreeMarker) {
    selectedOldTreeMarker.setIcon("/img/pin_29px.png");
    selectedOldTreeMarker.setZIndex(0);
  }

  if (selectedTreeMarker) {
    selectedTreeMarker.setIcon("/img/pin_32px.png");
    selectedTreeMarker.setZIndex(google.maps.Marker.MAX_ZINDEX);
  }
}

// using an index, get the point and marker and show them
function showMarkerInfoByIndex(index) {
  var point = points[index];
  var marker = markerByPointId[point["id"]];
  showMarkerInfo(point, marker, index);
}

// handle the index for a circular list
function getCircularPointIndex(index) {
  if (index > points.length - 1) {
    index = 0;
  } else if (index < 0) {
    index = points.length - 1;
  }
  return index;
}

// clear the markers from the map and then clear our the array of markers
function clearOverlays(overlays) {
  //console.log(overlays);
  for (var i = 0; i < overlays.length; i++) {
    //console.log(i);
    overlays[i].setMap(null);
  }
  overlays.length = 0;
}

// Gets the value of a given querystring in the provided url
function getQueryStringValue(name, url) {
  if (!url) url = window.location.href;
  name = name.replace(/[\[\]]/g, "\\$&");
  var regex = new RegExp("[?&]" + name + "(=([^&#]*)|&|#|$)"),
    results = regex.exec(url);
  if (!results) return null;
  if (!results[2]) return "";
  return decodeURIComponent(results[2].replace(/\+/g, " "));
}

function getPathVariable(name, url) {
  if (!url) url = window.location.href;
  console.log(url);
  var regex = new RegExp("/" + name + "/(.*)");
  console.log(regex);
  results = regex.exec(url);
  console.log(results);
  if (!results) return null;
  if (!results[1]) return "";
  return results[1];
}

function getHandleVariable(name, url) {
  if (!url) url = window.location.href;
  console.log(url);
  var regex = new RegExp("/@(.*)");
  console.log(regex);
  results = regex.exec(url);
  console.log(results);
  if (!results) return null;
  if (!results[1]) return "";
  return results[1];
}

// Returns the bounds for the visible area of the map.
// The offset parameter extends the bounds resulting rectangle by a certain percentage.
// For example: 1.1 will return a rectangle with each point (N, S, E, W) 10% farther from the rectangle.
// The offset specification might be useful for preloading trees around the visible area, taking advantage of a single request.
function getViewportBounds(offset) {
  var bounds = map.getBounds();
  if (offset) {
    offset -= 1;
    var east = bounds.getNorthEast().lng();
    var west = bounds.getSouthWest().lng();
    var north = bounds.getNorthEast().lat();
    var south = bounds.getSouthWest().lat();
    // Get the longitude and latitude differences
    var longitudeDifference = (east - west) * offset;
    var latitudeDifference = (north - south) * offset;

    // Move each point farther outside the rectangle
    // To west
    bounds.extend(new google.maps.LatLng(south, west - longitudeDifference));
    // To east
    bounds.extend(new google.maps.LatLng(north, east + longitudeDifference));
    // To south
    bounds.extend(new google.maps.LatLng(south - latitudeDifference, west));
    // To north
    bounds.extend(new google.maps.LatLng(north + latitudeDifference, east));
  }
  return bounds;
}

function toUrlValueLonLat(bounds) {
  var east = bounds.getNorthEast().lng();
  var west = bounds.getSouthWest().lng();
  var north = bounds.getNorthEast().lat();
  var south = bounds.getSouthWest().lat();
  return [east, north, west, south].join();
}

function determineInitialSize(latLng) {
  if (firstRender) {
    console.log("extends initial bounds");
    initialBounds.extend(latLng);
  }
}

function getClusterRadius(zoom) {
  switch (zoom) {
    case 1:
      return 10;
    case 2:
      return 8;
    case 3:
      return 6;
    case 4:
      return 4;
    case 5:
      return 0.8;
    case 6:
      return 0.75;
    case 7:
      return 0.3;
    case 8:
      return 0.099;
    case 9:
      return 0.095;
    case 10:
      return 0.05;
    case 11:
      return 0.03;
    case 12:
      return 0.02;
    case 13:
      return 0.008;
    case 14:
      return 0.005;
    case 15:
      return 0.004;
    case 16:
      return 0.003;
    case 17:
    case 18:
    case 19:
      return 0.0;
    default:
      return 0;
  }
}

function shortenLargeNumber(number) {
  var units = ["K", "M"],
    decimal;

  for (var i = units.length - 1; i >= 0; i--) {
    decimal = Math.pow(1000, i + 1);

    if (number <= -decimal || number >= decimal) {
      return +(number / decimal).toFixed(0) + units[i];
    }
  }
  return number;
}

function fitMapToBoundsForSet(data){
  const bounds = mapTools.getInitialBounds(
    data.map(i => {
      if(i.type === "cluster"){
        const c = JSON.parse(i.centroid);
        return {
          lat: c.coordinates[1],
          lng: c.coordinates[0],
        };
      }else if(i.type === "point"){
        return {
          lat: i.lat,
          lng: i.lon,
        };
      }
    }),
    window.innerWidth,
    window.innerHeight,
  );
  map.panTo(bounds.center);
  map.setZoom(bounds.zoomLevel);
}
 

//Initialize Google Maps and Marker Clusterer
var initialize = function() {
  console.log(window.location.href);
  token = getQueryStringValue("token") || null;
  organization = getQueryStringValue("organization") || null;
  treeid = getQueryStringValue("treeid") || null;
  userid = getQueryStringValue("userid") || null;
  flavor = getQueryStringValue("flavor") || null;
  donor = getQueryStringValue("donor") || null;
  wallet = getQueryStringValue("wallet") || null;
  if (wallet == null) {
    wallet = getHandleVariable("wallet") || null;
  }
  console.log(wallet);
  loader = document.getElementById("map-loader");

  var initialZoom = 2;
  var minZoom = 2;

  var linkZoom = parseInt(getQueryStringValue("zoom"));
  if (linkZoom) {
    initialZoom = linkZoom;
  }

//Fri Jul 17 14:26:03 CST 2020  do not set initial zoom when there is some parameters
//  if (
//    token != null ||
//    organization != null ||
//    treeid != null ||
//    userid !== null ||
//    donor != null
//  ) {
//    initialZoom = 10;
//    minZoom = null; // use the minimum zoom from the current map type
//  }

  var mapOptions = {
    zoom: initialZoom,
    minZoom: minZoom,
    mapTypeId: "hybrid",
    mapTypeControl: false,
    streetViewControl: false,
    fullscreenControl: false
  };

  console.log(mapOptions);

  map = new google.maps.Map(document.getElementById("map-canvas"), mapOptions);

  // only fetch when the user has made some sort of action
  google.maps.event.addListener(map, "dragstart", function() {
    fetchMarkers = true;
    firstInteraction = true;
  });

  function registerFirstInteraction() {
    firstInteraction = true;
  }

  google.maps.event.addListener(map, "click", registerFirstInteraction);

  google.maps.event.addListener(map, "mousemove", registerFirstInteraction);

  google.maps.event.addListener(map, "zoom_changed", function() {
    fetchMarkers = true;
  });

  /*
  google.maps.event.addListener(map, "idle", function() {
    console.log("triger idle...");
    var zoomLevel = !firstInteraction ? initialZoom : map.getZoom();
    console.log("New zoom level: " + zoomLevel);
    currentZoom = zoomLevel;
    initMarkers(toUrlValueLonLat(getViewportBounds(1.1)), zoomLevel);
  });
  */

//Fri Jul 17 14:26:56 CST 2020  do not use titlesloaded to set initial bounds
//use the firstRender in initMarkers fn to load initial bounds
//  // Adjust map bounds after it’s fully loaded, but only before first interaction
//  google.maps.event.addListener(map, "tilesloaded", function() {
//    if (
//      !firstInteraction &&
//      (token != null ||
//        organization != null ||
//        treeid != null ||
//        userid !== null ||
//        donor != null)
//    ) {
//      console.log("before first interaction, fit the map", initialBounds.toJSON());
//      map.fitBounds(initialBounds);
//    }
//  });


  google.maps.event.addListener(map, "idle", function() {
    console.log('IDLE');
    if(firstQuery){
      firstQuery = false
      let treeQueryParameters = getTreeQueryParametersFromRequestedFilters();
      if(treeQueryParameters == ""){
        var zoomLevel = !firstInteraction ? initialZoom : map.getZoom();
        console.log("New zoom level: " + zoomLevel);
        currentZoom = zoomLevel;
        initMarkers(toUrlValueLonLat(getViewportBounds(1.1)), zoomLevel);
        return 
      }


      let queryZoomLevel = 10;
      var clusterRadius = getQueryStringValue("clusterRadius") || getClusterRadius(queryZoomLevel);

      console.log("Cluster radius: " + clusterRadius);
      if (req != null) {
        console.log("initMarkers abort");
        req.abort();
      }
      var queryUrl = treetrackerApiUrl + "trees?clusterRadius=" + clusterRadius;
      queryUrl = queryUrl + "&zoom_level=" + queryZoomLevel;
      queryUrl = queryUrl + treeQueryParameters;

      console.log("request:", queryUrl);
      req = $.get(queryUrl, function(result) {

        if(result.data.length == 1){
          // rerun at higher zoom level
          let queryZoomLevel12 = 12
          var queryClusterRadius12 = getQueryStringValue("clusterRadius") || getClusterRadius(queryZoomLevel12);
          var queryUrl = treetrackerApiUrl + "trees?clusterRadius=" + queryClusterRadius12;
          queryUrl = queryUrl + "&zoom_level=" + queryZoomLevel12;
          queryUrl = queryUrl + treeQueryParameters;
          req = $.get(queryUrl, function(result) {
            fitMapToBoundsForSet(result.data)
          });

        } else {
          fitMapToBoundsForSet(result.data)
        }


      })

    } else {
      var zoomLevel = map.getZoom();
      console.log("New zoom level: " + zoomLevel);
      currentZoom = zoomLevel;
      initMarkers(toUrlValueLonLat(getViewportBounds(1.1)), zoomLevel);
    }
  });

  currentZoom = initialZoom;
  //map.setCenter({ lat: -3.33313276473463, lng: 37.142856230615735 });
  map.setCenter({ lat: 20, lng: 0 });

  $("#close-button").click(function() {
    $("#tree_info_div").hide("slide", "swing", 600);
    treeInfoDivShowing = false;
    $("#map-canvas").css("margin-left", "0px");
  });

  //initialize MapModel
  console.log("MAKING MAP MODELL");
  mapModel = new MapModel(treetrackerApiUrl);
  mapModel.map = map;
  mapModel.markers = markers;
};



google.maps.event.addDomListener(window, "load", initialize);
